File: FlexServ Backend/SpringBoot/BusinessService/src/main/java/com/flexserv/service/UserServiceImpl.java
package com.flexserv.service;

import java.util.List;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.flexserv.dto.request.AddressRequest;
import com.flexserv.dto.request.UpdateUserRequest;
import com.flexserv.dto.response.AddressResponse;
import com.flexserv.dto.response.UserResponse;
import com.flexserv.entity.Address;
import com.flexserv.entity.User;
import com.flexserv.exception.ResourceNotFoundException;
import com.flexserv.exception.UserAlreadyExistsException;
import com.flexserv.repository.AddressRepository;
import com.flexserv.repository.UserRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class UserServiceImpl implements UserService {

    private final UserRepository userRepository;
    private final AddressRepository addressRepository;

    @Override
    public UserResponse getUserProfile(Long userId) {
        User user = findUserById(userId);
        return mapToUserResponse(user);
    }

    @Override
    @Transactional
    public UserResponse updateUserProfile(Long userId, UpdateUserRequest request) {
        User user = findUserById(userId);

        // If phone is changed, check for uniqueness
        if (!user.getPhone().equals(request.getPhone()) && userRepository.existsByPhone(request.getPhone())) {
            throw new UserAlreadyExistsException("Phone number already in use by another account");
        }

        user.setName(request.getName());
        user.setPhone(request.getPhone());

        User updatedUser = userRepository.save(user);
        return mapToUserResponse(updatedUser);
    }

    @Override
    @Transactional
    public void deleteUserAccount(Long userId) {
        User user = findUserById(userId);
        userRepository.delete(user);
    }

    // --- Address CRUD Operations ---

    @Override
    @Transactional
    public AddressResponse addAddress(Long userId, AddressRequest request) {
        User user = findUserById(userId);

        Address address = new Address();
        address.setHouseNo(request.getHouseNo());
        address.setStreet(request.getStreet());
        address.setArea(request.getArea());
        address.setCity(request.getCity());
        address.setState(request.getState());
        address.setPincode(request.getPincode());

        user.addAddress(address); // Maintains bidirectional JPA relation
        Address savedAddress = addressRepository.save(address);

        return mapToAddressResponse(savedAddress);
    }

    @Override
    public List<AddressResponse> getUserAddresses(Long userId) {
        findUserById(userId); // Validates user existence
        List<Address> addresses = addressRepository.findByUserId(userId);
        return addresses.stream()
                .map(this::mapToAddressResponse)
                .collect(Collectors.toList());
    }

    @Override
    @Transactional
    public AddressResponse updateAddress(Long userId, Long addressId, AddressRequest request) {
        Address address = addressRepository.findByIdAndUserId(addressId, userId)
                .orElseThrow(() -> new ResourceNotFoundException("Address not found or doesn't belong to this user"));

        address.setHouseNo(request.getHouseNo());
        address.setStreet(request.getStreet());
        address.setArea(request.getArea());
        address.setCity(request.getCity());
        address.setState(request.getState());
        address.setPincode(request.getPincode());

        Address updatedAddress = addressRepository.save(address);
        return mapToAddressResponse(updatedAddress);
    }

    @Override
    @Transactional
    public void deleteAddress(Long userId, Long addressId) {
        User user = findUserById(userId);
        Address address = addressRepository.findByIdAndUserId(addressId, userId)
                .orElseThrow(() -> new ResourceNotFoundException("Address not found or doesn't belong to user"));

        user.removeAddress(address);
        addressRepository.delete(address);
    }

    // --- Helper Mapping Methods ---

    private User findUserById(Long userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + userId));
    }

    private UserResponse mapToUserResponse(User user) {
        UserResponse response = new UserResponse();
        response.setId(user.getId());
        response.setName(user.getName());
        response.setEmail(user.getEmail());
        response.setPhone(user.getPhone());
        response.setRole(user.getRole());
        response.setActive(user.getIsActive());
        response.setCreatedAt(user.getCreatedAt());
        return response;
    }

    private AddressResponse mapToAddressResponse(Address address) {
        return new AddressResponse(
                address.getId(),
                address.getHouseNo(),
                address.getStreet(),
                address.getArea(),
                address.getCity(),
                address.getState(),
                address.getPincode()
        );
    }
}

File: FlexServ Backend/DotNet/AdminService/Services/UserServiceImpl.cs
using AdminService.Data;
using AdminService.Dtos;
using AdminService.Models;
using System;
using System.Collections.Generic;
using System.Linq;

namespace AdminService.Services
{
    public class UserServiceImpl : IUserService
    {
        private readonly AdminDbContext _dbContext;

        public UserServiceImpl(AdminDbContext dbContext)
        {
            _dbContext = dbContext;
        }

        public UserResponse GetUserProfile(long userId)
        {
            var user = FindUserById(userId);
            return MapToUserResponse(user);
        }

        public UserResponse UpdateUserProfile(long userId, UpdateUserRequest request)
        {
            var user = FindUserById(userId);

            if (user.Phone != request.Phone && _dbContext.Users.Any(u => u.Phone == request.Phone))
            {
                throw new InvalidOperationException("Phone number already in use by another account");
            }

            user.Name = request.Name;
            user.Phone = request.Phone;
            user.UpdatedAt = DateTime.UtcNow;

            _dbContext.SaveChanges();
            return MapToUserResponse(user);
        }

        public void DeleteUserAccount(long userId)
        {
            var user = FindUserById(userId);
            _dbContext.Users.Remove(user);
            _dbContext.SaveChanges();
        }

        public AddressResponse AddAddress(long userId, AddressRequest request)
        {
            var user = FindUserById(userId);

            var address = new Address
            {
                UserId = user.Id,
                HouseNo = request.HouseNo,
                Street = request.Street,
                Area = request.Area,
                City = request.City,
                State = request.State,
                Pincode = request.Pincode
            };

            _dbContext.Addresses.Add(address);
            _dbContext.SaveChanges();

            return MapToAddressResponse(address);
        }

        public List<AddressResponse> GetUserAddresses(long userId)
        {
            FindUserById(userId);
            var addresses = _dbContext.Addresses.Where(a => a.UserId == userId).ToList();
            return addresses.Select(a => MapToAddressResponse(a)).ToList();
        }

        public AddressResponse UpdateAddress(long userId, long addressId, AddressRequest request)
        {
            var address = _dbContext.Addresses.FirstOrDefault(a => a.Id == addressId && a.UserId == userId)
                ?? throw new KeyNotFoundException("Address not found or doesn't belong to this user");

            address.HouseNo = request.HouseNo;
            address.Street = request.Street;
            address.Area = request.Area;
            address.City = request.City;
            address.State = request.State;
            address.Pincode = request.Pincode;

            _dbContext.SaveChanges();
            return MapToAddressResponse(address);
        }

        public void DeleteAddress(long userId, long addressId)
        {
            FindUserById(userId);
            var address = _dbContext.Addresses.FirstOrDefault(a => a.Id == addressId && a.UserId == userId)
                ?? throw new KeyNotFoundException("Address not found or doesn't belong to user");

            _dbContext.Addresses.Remove(address);
            _dbContext.SaveChanges();
        }

        private User FindUserById(long userId)
        {
            return _dbContext.Users.FirstOrDefault(u => u.Id == userId)
                ?? throw new KeyNotFoundException($"User not found with id: {userId}");
        }

        private static UserResponse MapToUserResponse(User user)
        {
            return new UserResponse
            {
                Id = user.Id,
                Name = user.Name,
                Email = user.Email,
                Phone = user.Phone,
                Role = user.Role,
                Active = user.IsActive,
                CreatedAt = user.CreatedAt
            };
        }

        private static AddressResponse MapToAddressResponse(Address address)
        {
            return new AddressResponse(
                address.Id,
                address.HouseNo,
                address.Street,
                address.Area,
                address.City,
                address.State,
                address.Pincode
            );
        }
    }
}
